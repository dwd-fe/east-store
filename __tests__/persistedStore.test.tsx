import { createStore } from '../'
import React from 'react'

import renderer, { act, ReactTestRenderer } from 'react-test-renderer'

describe('persistedStore', () => {
  test('persist store after unmount', () => {
    const persistedStore = createStore(
      0,
      {
        increase: () => count => count + 1,
        decrease: n => count => count - n
      },
      { persist: true }
    )

    const Com1: React.FC = () => {
      const [count, action] = persistedStore.useState()
      const handleDecrease = () => {
        action.decrease(3)
      }
      return (
        <div>
          <span>{count}</span>
          <button id="increase-btn" onClick={action.increase}>
            increase
          </button>
          <button id="decrease-btn" onClick={handleDecrease}>
            decrease
          </button>
        </div>
      )
    }

    const Com2: React.FC = () => {
      const [count, _] = persistedStore.useState()
      return (
        <div>
          <span id="count">{count}</span>
        </div>
      )
    }

    const App: React.FC = () => {
      return (
        <div>
          <Com1 />
          <Com2 />
        </div>
      )
    }

    let component = {} as ReactTestRenderer
    act(() => {
      component = renderer.create(<App />)
    })
    expect(component.toJSON()).toMatchSnapshot()

    act(component.root.findByProps({ id: 'increase-btn' }).props.onClick)
    // shared
    expect(component.root.findByProps({ id: 'count' }).children[0]).toBe('1')
    expect(component.toJSON()).toMatchSnapshot()

    component.unmount()

    // recreate
    act(() => {
      component = renderer.create(<App />)
    })
    expect(component.root.findByProps({ id: 'count' }).children[0]).toBe('1')
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('use custom storage', () => {
    const m = new Map()
    const g = jest.fn((name: string): string => {
      return name
    })
    const set = jest.fn((key: string, val: number) => {
      return m.set(key, val)
    })
    const get = jest.fn((key: string): number => {
      return m.get(key)
    })
    const myStorage = {
      generateKey: g,
      set,
      get
    }
    const NAME = 'FooStore'
    const store = createStore(
      0,
      {
        increase: () => count => count + 1
      },
      { persist: myStorage, name: NAME }
    )

    const Counter: React.FC = () => {
      const [count, action] = store.useState()
      return (
        <div>
          <span id="count">{count}</span>
          <button id="increase-btn" onClick={action.increase}>
            increase
          </button>
        </div>
      )
    }

    let component = {} as ReactTestRenderer
    act(() => {
      component = renderer.create(<Counter />)
    })
    expect(component.toJSON()).toMatchSnapshot()

    expect(g).toBeCalledWith(NAME)
    expect(get).toBeCalledWith(NAME)

    act(component.root.findByProps({ id: 'increase-btn' }).props.onClick)

    expect(set).toBeCalledTimes(0)
    component.unmount()
    expect(set).toBeCalledWith(NAME, 1)
  })
})
