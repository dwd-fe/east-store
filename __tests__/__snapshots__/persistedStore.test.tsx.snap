// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`persistedStore persist store after unmount 1`] = `
<div>
  <div>
    <span>
      0
    </span>
    <button
      id="increase-btn"
      onClick={[Function]}
    >
      increase
    </button>
    <button
      id="decrease-btn"
      onClick={[Function]}
    >
      decrease
    </button>
  </div>
  <div>
    <span
      id="count"
    >
      0
    </span>
  </div>
</div>
`;

exports[`persistedStore persist store after unmount 2`] = `
<div>
  <div>
    <span>
      1
    </span>
    <button
      id="increase-btn"
      onClick={[Function]}
    >
      increase
    </button>
    <button
      id="decrease-btn"
      onClick={[Function]}
    >
      decrease
    </button>
  </div>
  <div>
    <span
      id="count"
    >
      1
    </span>
  </div>
</div>
`;

exports[`persistedStore persist store after unmount 3`] = `
<div>
  <div>
    <span>
      1
    </span>
    <button
      id="increase-btn"
      onClick={[Function]}
    >
      increase
    </button>
    <button
      id="decrease-btn"
      onClick={[Function]}
    >
      decrease
    </button>
  </div>
  <div>
    <span
      id="count"
    >
      1
    </span>
  </div>
</div>
`;

exports[`persistedStore use custom storage 1`] = `
<div>
  <span
    id="count"
  >
    0
  </span>
  <button
    id="increase-btn"
    onClick={[Function]}
  >
    increase
  </button>
</div>
`;
